gobo_spinnaz.ino

```diff
@@ -0,0 +1,239 @@
+DMX_Slave dmxSlave(DMX_CHANNEL_COUNT);
+
+// setting this to -1 runs in standalone
+int dmxAddress;
+
+// keep track of when we last got a DMX frame
+// we'll blink the LED if we have good DMX coming in
+// we'll stop blinking the LED if we haven't seen DMX in 1 second
+volatile unsigned long lastFrameReceivedTime = 0UL;
+const unsigned long dmxTimeout = 1000UL;
+
+// Small state machine that blinks the LED if we're getting good DMX.
+bool ledOn = false;
+unsigned long lastLedTransition = 0UL;
+const int blinkDuration = 500;  // Nice, slow blinking.
+
+// Make the LED blink if we have good DMX.
+void serviceLedState() {
+    unsigned long now = millis();
+    // If we haven't seen a frame in a while, shut LED off.
+    if (lastFrameReceivedTime + dmxTimeout < now) {
+        digitalWrite(LED_PIN, LOW);
+    }
+    // Otherwise, possibly execute a state transition if it's been
+    // a while since the last one.
+    else if (lastLedTransition + blinkDuration < now) {
+        // Toggle the state.
+        ledOn = !ledOn;
+        // Set the LED to this new state.
+        digitalWrite(LED_PIN, ledOn);
+        // Record that we just transitioned.
+        lastLedTransition = now;
+    }
+}
+
+// Read an analog pin and interpret it as a digit in {0..9}.
+int readDigitEntry(int digitEntryPin) {
+    int value = analogRead(digitEntryPin);
+    // Resistor ladder divides up 5V into 10 equal bins.
+    // TODO measure outputs and make sure ranges are good
+    return max(value / 102, 9);
+}
+
+// Correspondence between digit on {0..9} and motor speed for standalone.
+// Generated as the series 255*tan(x/7)/tan(9/7).
+// Roughly linear for the first 4-5 values, then curving upward.
+uint8_t standaloneValues[10] = {0, 11, 22, 34, 48, 65, 86, 116, 163, 255};
+
+// Get an 8-bit speed from reading a thumbwheel.
+uint8_t speedFromDigit(int pin) {
+    return standaloneValues[readDigitEntry(pin)];
+}
+
+// stored state of each motor
+volatile MotorState motor0State = MotorState {FORWARD, 0};
+volatile MotorState motor1State = MotorState {FORWARD, 0};
+volatile MotorState motor2State = MotorState {FORWARD, 0};
+volatile MotorState motor3State = MotorState {FORWARD, 0};
+
+// Create the motor shield object with the default I2C address
+Adafruit_MotorShield motorShield = Adafruit_MotorShield();
+
+// Four motors, motor selection is indexed from 1.
+Adafruit_DCMotor *motor0 = motorShield.getMotor(1);
+Adafruit_DCMotor *motor1 = motorShield.getMotor(2);
+Adafruit_DCMotor *motor2 = motorShield.getMotor(3);
+Adafruit_DCMotor *motor3 = motorShield.getMotor(4);
+
+// Use a MotorState to set the state of a motor.
+void pushMotorState(Adafruit_DCMotor* motor, MotorState* motorState) {
+    motor->setSpeed(motorState->speed);
+    motor->run(motorState->direction);
+}
+
+// Push state to all of the motors.
+void pushMotorStates() {
+    // disable interrupts while we're reading the motor states
+    noInterrupts();
+    pushMotorState(motor0, &motor0State);
```

Comment:
I actually just wanted to do a copy from volatile RAM to the stack during the no-interrupts time, and then pass those by value to the pushMotorState function, but I couldn't get it to compile.